server/src/tls.rs

```diff
@@ -0,0 +1,360 @@
+    }
+
+    // create verifier : does not allow unauthenticated clients
+    // and authenticated clients must be certified by one of the listed CAs
+    let client_cert_verifier = rustls::server::AllowAnyAuthenticatedClient::new(client_auth_roots).boxed();
+
+    // make config
+    let mut config : ServerConfig = ServerConfig::builder()
+        // Allow everything available in rustls for maximum support
+        .with_cipher_suites(rustls::ALL_CIPHER_SUITES)
+        .with_safe_default_kx_groups()
+        .with_protocol_versions(rustls::ALL_VERSIONS)
+        .expect("Could not build configuration defaults")
+        .with_client_cert_verifier(client_cert_verifier)  // add verifier
+        .with_single_cert(cert, priv_key)                 // add server vertification
+        .expect("Bad configuration certificate or key");
+
+    // any http version is ok
+    config.alpn_protocols = vec![b"h2".to_vec(), b"http/1.1".to_vec(), b"http/1.0".to_vec()];
+
+    info!("Loaded TLS configuration with server certificate {}", args.server_certificate());
+
+    Ok(ResultConfig {server_config: Arc::new(config), thumbprint})
+}
+
+
+/// Get machine name from certificate if subject is non-empty
+///
+/// Panics
+///
+/// - When certificate could not be parsed (error on x509_parser::X509Certificate::from_der)
+pub fn subject_from_cert(cert : &[u8]) -> Option<String>
+{
+    // load certificate to decompose its content
+    let cert = X509Certificate::from_der(&cert).unwrap().1;
+
+    let oid_registry = OidRegistry::default().with_x509(); // registry of OIDs we will need
+    let rdn_iter = cert.subject.iter_rdn(); // iterator on RDNs
+
+    for subject_attribute in rdn_iter
+    {
+        // Each entry contains a list of AttributeTypeAndValue objects,
+        // so we fetch their attribute type (= the OID representing each of them)
+        let sn = subject_attribute.iter();
+
+        for set in sn
+        {
+            // OID of the sub-entry
+            let typ = set.attr_type();
+
+            // we match the OID with its signification (SN)
+            let oid_reg = match oid_registry.get(typ)
+            {
+                Some(oid) => Some(oid.sn()),  // get the SN corresponding to the OID
+                None => None
+            };
+
+            // the value we are interested in is only contained where the commonName is
+            if oid_reg == Some("commonName") {
+                // get data as text => FQDN of the client
+                match set.as_str() {
+                    Ok(name) => return Some(name.to_string()),
+                    Err(_) => return None
+                };
+            }
+        }
+    }
+    None
+}
+
+
+/// Read and decode request payload
+pub async fn get_request_payload(
+    settings: &Collector,
+    req: Request<Body>,
+) -> Result<Option<String>> {
+    let (parts, body) = req.into_parts();
+
+    let response_content_length = body
```

Comment:
This function is roughly a copy/paste of `kerberos::get_request_payload` (at least until line 221). Could you factorize it?